descriptors/web_app.js

```diff
@@ -0,0 +1,199 @@
+            'descriptor characteristic mismatch');
+          assertEquals(null, descriptor.value, 'value should be null');
+          const value = await descriptor.readValue();
+          const decoder = new TextDecoder('utf-8');
+          let strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          strVal = decoder.decode(descriptor.value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          const encoder = new TextEncoder('utf-8');
+          try {
+            await descriptor.writeValue(encoder.encode('Something else'));
+            logError('Expected exception');
+          } catch (e) {
+            assertEquals('NotSupportedError', e.name);
+          }
+          break;
+        case characteristicWithoutDescriptorUUID:
+          try {
+            await characteristic.getDescriptors();
+            logError(`Characteristic ${characteristic.uuid} should have no descriptors`);
+          } catch (e) {
+            assertEquals('NotFoundError', e.name);
+          }
+          break;
+        default:
+          logError(`Unexpected characteristic: ${characteristic.uuid}`);
+      }
+    }
+  }
+
+  // Verify descriptors for Nordic UART Service (NUS)
+  // https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bluetooth/services/nus.html
+  const verifyNordicServiceDescriptors = async (service) => {
+    assertEquals(nordicUARTServiceUUID, service.uuid, 'incorrect service UUID');
+
+    const characteristics = await service.getCharacteristics();
+    assertEquals(2, characteristics.length, 'expected 2 characteristic');
+    for (characteristic of characteristics) {
+      switch (characteristic.uuid) {
+        case '6e400002-b5a3-f393-e0a9-e50e24dcca9e': // RX
+          try {
+            await characteristic.getDescriptors();
+            throw 'RX Characteristic should have no descriptors';
+          } catch (e) {
+            assertEquals('NotFoundError', e.name);
+          }
+          break;
+        case '6e400003-b5a3-f393-e0a9-e50e24dcca9e': // TX
+          const descriptors = await characteristic.getDescriptors();
+          assertEquals(1, descriptors.length);
+          const descriptor = descriptors[0];
+          assertEquals(
+            BluetoothUUID.getDescriptor('gatt.client_characteristic_configuration'),
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+
+          // Not currently subscribed to notifications, so expect notifications
+          // disabled in the descriptor.
+          let dataView = await descriptor.readValue();
+          value = dataView.getUint16(0, /*littleEndian=*/true);
+          // CCCD (Bluetooth 4.2) only uses bottom two bits. We expect to see
+          // LSB zero to signify nontifications disabled.
+          assertEquals(0x0, (value & 0x3));
```

Comment:
Is this masking with 0x3 so that it also passes if we set the indications bit instead? If that's intentional mention it in the comment.